developer/python-django-devtools-reference.rst

```diff
@@ -0,0 +1,229 @@
+on `its pypi page <https://pypi.org/project/ipdb/>`_.
+
+How to enable these dev tools in OpenWISP modules
+-------------------------------------------------
+
+These tools can be implemented in OpenWISP modules and can significantly
+improve the workflow. Let's do a walkthrough of how to use them in
+`OpenWISP Controller <https://github.com/openwisp/openwisp-controller>`_
+as an example. In the ``tests/`` folder, ``local_settings_example.py``
+must be copied and renamed to ``local_settings.py`` as all the changes
+to be made or code to be added will be made and added there. This
+technique can be used in any other openwisp module too.
+
+.. code-block:: bash
+
+	cd tests/
+	cp local_settings_example.py local_settings.py
+
+To start, the `steps <https://github.com/openwisp/openwisp-controller>`_
+for installing OpenWISP Controller must be followed. The commands
+``pipenv run ./manage.py migrate`` and
+``pipenv run ./manage.py createsuperuser`` must be ran and
+``SPATIALITE_LIBRARY_PATH`` should be specified in the
+``local_settings.py`` file.
+
+Coming to the tools, instead of ``python manage.py runserver``, the
+command ``python manage.py runserver_plus`` can be used which will
+provide more debugging information. It will show all the processes
+occuring on the locally hosted website.
+
+Also, ``ipython`` can be used along ``shell_plus`` by running the command
+``./manage.py shell_plus --ipython`` in the terminal. This will provide
+an interactive shell for running code in python.
+
+To debug the code, ``ipdb`` can be used. Commands similar to
+``ipdb mymodule.py`` may be used to carry out that process. A
+list of lines where errors have been found or lines which can be further
+optimized will be returned.
+
+Lastly, ``django-debug-toolbar`` may be used along to display information
+about processes occuring on the website. To achieve that, some code needs
+to be added to our current module, i.e ``openwisp-controller``. To begin,
+the lines ``'debug_toolbar'`` and
+``'debug_toolbar.middleware.DebugToolbarMiddleware'`` need to be added
+to the ``INSTALLED_APPS`` and to the``MIDDLEWARE`` settings respectively
+and the line ``INTERNAL_IPS = ('127.0.0.1')`` should be mentioned else
+the django debug toolbar won't be displayed. Here's what
+``local_settings.py`` will roughly look like at the end:
+
+.. code-block:: python
+
+	from django.conf import settings
+
+	settings.INSTALLED_APPS += [
+	    'debug_toolbar'
+	]
+
+	SPATIALITE_LIBRARY_PATH = 'mod_spatialite'
+
+	settings.MIDDLEWARE.append\
+	('debug_toolbar.middleware.DebugToolbarMiddleware')
+
+	INTERNAL_IPS = ('127.0.0.1')
+
+To complete the process, the Debug Toolbar’s URL needs to be added
+to the URLconf of ``openwisp-controller`` as shown in `this
+<django-debug-toolbar.readthedocs.io/en/latest/installation.html>`_
+tutorial but this is already present in the last lines of ``urls.py`` so
+this step can be ignored. Now, logging to 127.0.0.1:8000 and inputting
```

Comment:
add a link to show how it's done: https://github.com/openwisp/openwisp-controller/blob/master/tests/urls.py#L23-L25